collector/inode_linux.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !noinode
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+)
+
+const (
+	inodeSubsystem = "inode"
+)
+
+type inodeCollector struct {
+	metrics map[string]prometheus.Gauge
+}
+
+func init() {
+	Factories[inodeSubsystem] = NewInodeCollector
+}
+
+// NewInodeCollector returns a new Collector exposing inode-state stats.
+func NewInodeCollector() (Collector, error) {
+	return &inodeCollector{
+		metrics: map[string]prometheus.Gauge{},
+	}, nil
+}
+
+func (c *inodeCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	inode, err := getInodeStats(procFilePath("sys/fs/inode-state"))
+	if err != nil {
+		return fmt.Errorf("couldn't get inode-state-nr: %s", err)
+	}
+	for name, value := range inode {
+		if _, ok := c.metrics[name]; !ok {
+			c.metrics[name] = prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Namespace: Namespace,
+					Subsystem: inodeSubsystem,
+					Name:      name,
+					Help:      fmt.Sprintf("Inode statistics: %s.", name),
+				},
+			)
+		}
+		v, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return fmt.Errorf("invalid value %s in inode-state: %s", value, err)
+		}
+		c.metrics[name].Set(v)
+	}
+	for _, m := range c.metrics {
+		m.Collect(ch)
+	}
+	return err
+}
+
+func getInodeStats(fileName string) (map[string]string, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	return parseInodeStats(file, fileName)
+}
+
+func parseInodeStats(r io.Reader, fileName string) (map[string]string, error) {
+	var scanner = bufio.NewScanner(r)
```

Comment:
Using a scanner for file with a single line seems to be unnecessary. Just reading everything from the reader into a byte array should be enough. The error check of the scanner is also missing.